### PR TITLE
[RHOAIENG-47414] Fix CWE-918: Add SSRF protection to RemotePayloadProcessor

### PR DESCRIPTION
**Pull request by mholder6**
Fix CWE-918: Add SSRF protection to RemotePayloadProcessor
  - Added URI validation to prevent Server-Side Request Forgery (CWE-918)
  - Blocks requests to localhost, loopback, private IPs, and link-local addresses
  - Prevents access to cloud metadata services 
  - Added comprehensive tests for SSRF protection

### CODE REVIEW

File: src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java [hunk 1]
```diff
@@ -49,11 +51,73 @@ public class RemotePayloadProcessor implements PayloadProcessor {
...
+
+            // Additional check for 0.0.0.0
+            if (addr.isAnyLocalAddress()) {
+                throw new IllegalArgumentException("URI cannot point to wildcard address: " + host);
+            }
+
+        } catch (UnknownHostException e) {
+            // If the host can't be resolved, allow it to continue
+            // The actual HTTP request will fail later with a proper error
+            logger.debug("Unable to resolve host for validation: {}", host);
```
Comment (ruivieira) [inline src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java]: Since this is an error, log with error (or at least warning or info), instead of debug

File: src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java [hunk 2]
```diff
@@ -49,11 +51,73 @@ public class RemotePayloadProcessor implements PayloadProcessor {
...
+        if (uri == null) {
+            throw new IllegalArgumentException("URI cannot be null");
+        }
+
+        String host = uri.getHost();
+        if (host == null || host.isEmpty()) {
+            throw new IllegalArgumentException("URI must have a valid host");
+        }
+
+        // Reject localhost and loopback addresses (string-based checks)
```
Comment (ruivieira) [inline src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java]: Should we add more comprehensive checks for localhost bypass (eg `[::]`, `http://0177.0.0.1/`, etc)?

File: src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java [hunk 3]
```diff
@@ -49,11 +51,73 @@ public class RemotePayloadProcessor implements PayloadProcessor {
...
     private final HttpClient client;
 
+    /**
+     * Validates the URI to prevent SSRF attacks by rejecting requests to private/internal networks.
+     *
+     * @param uri The URI to validate
+     * @throws IllegalArgumentException if the URI points to a private or internal network
+     */
+    private static void validateUri(URI uri) {
+        if (uri == null) {
```
Comment (ruivieira) [inline src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java:61]: Should we also validate scheme (i.e. either http or https?)

File: src/test/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessorTest.java [hunk 4]
```diff
@@ -67,4 +68,60 @@ void testDestinationUnreachableHTTPS() throws IOException, NoSuchAlgorithmExcept
...
+        }, "Should reject 172.16-31.x.x private IP");
+    }
+
+    @Test
+    void testSSRFProtection_AWSMetadata() {
+        assertThrows(IllegalArgumentException.class, () -> {
+            new RemotePayloadProcessor(URI.create("http://169.254.169.254/latest/meta-data/"));
+        }, "Should reject AWS metadata service");
+    }
+
```
Comment (ruivieira) [inline src/test/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessorTest.java:120]: Should we add a test for empty host (`http://:8080`)?

Also for some of the localhost bypasses in https://github.com/red-hat-data-services/modelmesh/pull/940/changes#r2747578445

File: src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java [hunk 5]
```diff
@@ -49,11 +51,73 @@ public class RemotePayloadProcessor implements PayloadProcessor {
...
+            throw new IllegalArgumentException("URI cannot point to localhost or loopback address: " + host);
+        }
+
+        // Try to resolve the host and perform additional validation
+        // If the host can't be resolved, allow it to pass (the HTTP request will fail later)
+        try {
+            InetAddress addr = InetAddress.getByName(host);
+
+            // Reject private IP ranges (RFC 1918)
+            if (addr.isSiteLocalAddress()) {
```
Comment (ruivieira) [inline src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java]: Doesn't validate for `fc00::/7`, but not sure.

**Review by ruivieira — COMMENTED** [review]
@mholder6 thanks for the PR, I've left a few comments.

**Comment by spolti** [comment]
Hey @ruivieira, thanks for reviewing it, wondering which ip address we expect here, is it only external addresses? Assuming it is ok to block localhost and the other ones.

File: src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java (hunk 5 above)
Comment (mholder6) [reply src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java]: Added explicit validation at lines 131-134 using bitwise check `(addrBytes[0] & 0xfe) == 0xfc` to catch both` fc00::/8` and `fd00::/8` ranges. This now blocks IPv6 private addresses that `isSiteLocalAddress()` missed.

File: src/test/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessorTest.java (hunk 4 above)
Comment (mholder6) [reply src/test/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessorTest.java:120]: Added 11 new test cases covering:                                                   
  - Empty host (`http://:8080`)                                                       
  - Localhost bypasses (octal, hex, IPv6 variants)                                    
  - IPv6 ULA addresses                                                                
  - Invalid schemes

File: src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java (hunk 3 above)
Comment (mholder6) [reply src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java:61]: Added scheme validation at lines 66-69. Now only HTTP and HTTPS are allowed, rejecting file://, ftp://, gopher://, etc. Also added test coverage for invalid schemes.

File: src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java (hunk 2 above)
Comment (mholder6) [reply src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java]: Added comprehensive checks for:                                                     
  - IPv6 variants: `[::]`, `::`                                                       
  - Octal encoding: `0177.0.0.1` (regex pattern detection)                            
  - Hex encoding: `0x7f.0.0.1`                                                        
  - Wildcard: `0.0.0.0`                                                               
                                                                                      
  See lines 80-94 for string-based validation and corresponding test cases.

File: src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java (hunk 1 above)
Comment (mholder6) [reply src/main/java/com/ibm/watson/modelmesh/payload/RemotePayloadProcessor.java]: Changed to `logger.warn()` at line 140. This is now properly logged as a security validation event

**Comment by ruivieira** [comment]
> Hey @ruivieira, thanks for reviewing it, wondering which ip address we expect here, is it only external addresses? Assuming it is ok to block localhost and the other ones.

@spolti We only set this as a valid Kubernetes service, pointing to a `TrustyAIService`, and although there might be legitimate uses for `localhost` (for instance local testing), in general this would be an _external_ service consuming the payload-forwards.

**Review by spolti — APPROVED** [review]
